### PR TITLE
Draggable objects and decos

### PR DESCRIPTION
**Pull request by PlasmaFlare**
Working implementation for issue https://github.com/chicory-pizza/chicory-data/issues/37

- How it works:
   - Replace div that wraps around the object/deco with a `TransformDiv`
   - `TransformDiv` updates its own internal position when user drags the div around
   - When user stops dragging, `TransformDiv` emits event of updated position. `LevelInspector` picks up on the change and calls dispatch to update object/deco properties
- Note: Most of the code is designed in mind for supporting more transform operations in the future, such as scaling and rotating. But for now, moving is the only supported operation.

### CODE REVIEW

File: src/levelEditor/preview/TransformDiv.jsx [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+		});
+	}, [
+		baseTransform.x,
+		baseTransform.y,
+		baseTransform.xScale,
+		baseTransform.yScale,
+		baseTransform.angle,
+	]);
+
+	function transformReducer(
```
Comment (ngyikp) [inline src/levelEditor/preview/TransformDiv.jsx:87]: The reducer is probably overkill right now, but it makes sense when we want to add more functionality 👍 

Put the reducer function outside the component to be clearer

File: src/levelEditor/preview/TransformDiv.jsx [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+		};
+	}, [onMouseUp]);
+
+	// Mainly for updating transform when it gets updated by something other than dragging/dropping
+	useEffect(() => {
+		dispatch({
+			type: 'set',
+			transform: baseTransform,
+		});
+	}, [
```
Comment (ngyikp) [inline src/levelEditor/preview/TransformDiv.jsx:79]: The `baseTransform` object is being reconstructed on every re-render, listing out each property just masks the problem

File: src/levelEditor/preview/TransformDiv.jsx [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+			className={className}
+			onMouseDown={(ev) => onMouseDown(ev, 'move')}
+			style={getTransformStyle()}
+		>
+			{children}
+		</div>
+	);
+}
+
+export default (memo<Props>(TransformDiv): React$AbstractComponent<
```
Comment (ngyikp) [inline src/levelEditor/preview/TransformDiv.jsx:179]: memo is only useful if it's carefully optimized to reuse the same functions/objects on every re-render

Right now the components that are using `<TransformDiv>` aren't doing that, so memo will never work and waste CPU, let's remove it

File: src/levelEditor/preview/TransformDiv.jsx [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+	origin: ?[number, number],
+	onTransformUpdate: (transform: TransformType) => mixed,
+	children: React$Node,
+	...
+}>;
+
+type ReducerAction =
+	| {type: 'move', deltaX: number, deltaY: number}
+	| {type: 'set', transform: TransformType}
+	| {type: null};
```
Comment (ngyikp) [inline src/levelEditor/preview/TransformDiv.jsx:23]: This null action shouldn't be needed?

File: src/levelEditor/preview/TransformDiv.jsx [hunk 5]
```diff
@@ -0,0 +1,182 @@
+// @flow strict
+
+import {memo, useCallback, useEffect, useReducer, useState} from 'react';
+
+import type {TransformType} from '../types/TransformType';
+
+type TransformAction = 'move';
```
Comment (ngyikp) [inline src/levelEditor/preview/TransformDiv.jsx:7]:
```suggestion
type TransformAction = 'MOVE';
```

File: src/levelEditor/preview/TransformDiv.jsx [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+	onTransformUpdate,
+	origin,
+	className,
+	...otherProps
+}: Props): React$Node {
+	const [currentTransform, dispatch] = useReducer(
+		transformReducer,
+		baseTransform
+	);
+	const [transformAction, setTransformAction] =
```
Comment (ngyikp) [inline src/levelEditor/preview/TransformDiv.jsx:39]: To be clearer
```suggestion
	const [currentTransformAction, setCurrentTransformAction] =
```

**Review by ngyikp — CHANGES_REQUESTED** [review]
Let's also show the move cursor when moving?

File: src/levelEditor/preview/TransformDiv.jsx (hunk 1 above)
Comment (PlasmaFlare) [reply src/levelEditor/preview/TransformDiv.jsx:87]: By "outside", you mean move the function definition to be outside the `TransformDiv` function?

Comment (ngyikp) [reply src/levelEditor/preview/TransformDiv.jsx:87]: Yes

File: src/levelEditor/preview/TransformDiv.jsx (hunk 2 above)
Comment (PlasmaFlare) [reply src/levelEditor/preview/TransformDiv.jsx:79]: That still doesn't tell me what I should do.

File: src/levelEditor/preview/LevelPreviewObjects.jsx [hunk 7]
```diff
@@ -32,29 +42,19 @@ function LevelPreviewObjects(props: Props): React$Node {
...
+				IDLE_ORIGIN_X / DOG_RES_SCALE,
+				IDLE_ORIGIN_Y / DOG_RES_SCALE,
+			];
 		}
 
 		return (
 			// eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions
-			<div
+			<TransformDiv
+				baseTransform={{x: obj.x, y: obj.y}}
```
Comment (ngyikp) [inline src/levelEditor/preview/LevelPreviewObjects.jsx:57]: This breaks rotation/scaling for objCustomDog, but I'll fix them later as that's a rarely used feature

**Review by ngyikp — APPROVED** [review]

**Pull request merged**